app/src/main/java/com/twilio/video/app/sdk/RoomFactory.kt

```diff
@@ -0,0 +1,143 @@
+package com.twilio.video.app.sdk
+
+import android.content.Context
+import android.content.SharedPreferences
+import com.twilio.androidenv.Env
+import com.twilio.video.AudioCodec
+import com.twilio.video.BandwidthProfileMode
+import com.twilio.video.BandwidthProfileOptions
+import com.twilio.video.ConnectOptions
+import com.twilio.video.EncodingParameters
+import com.twilio.video.G722Codec
+import com.twilio.video.H264Codec
+import com.twilio.video.IsacCodec
+import com.twilio.video.NetworkQualityConfiguration
+import com.twilio.video.NetworkQualityVerbosity
+import com.twilio.video.OpusCodec
+import com.twilio.video.PcmaCodec
+import com.twilio.video.PcmuCodec
+import com.twilio.video.Room
+import com.twilio.video.Video
+import com.twilio.video.VideoBandwidthProfileOptions
+import com.twilio.video.VideoCodec
+import com.twilio.video.Vp8Codec
+import com.twilio.video.Vp9Codec
+import com.twilio.video.app.data.Preferences
+import com.twilio.video.app.data.api.TokenService
+import com.twilio.video.app.util.EnvUtil
+
+class RoomFactory(
```

Comment:
This may be too nit, but I think the pattern and naming here masks the API of the Video SDK a little too much. Ideally, I'd like developers to see the verbs/actions associated with the original API so the experience is similar even in the presence of reasonable code abstractions like this. What are your thoughts on renaming `RoomFactory` to `VideoClient` and `newInstance` to `connect`.

Another option would be to implement an extension method off the `Video` class that allows connecting with a co-routine context.

Reply:
Additionally, what are your thoughts on removing SharedPreferences from this class? It would be nice if this class did not know about shared preferences and simply accepted connect options as a method parameter. The call site would be responsible for getting the data for the connect options.

Reply:
> This may be too nit, but I think the pattern and naming here masks the API of the Video SDK a little too much. Ideally, I'd like developers to see the verbs/actions associated with the original API so the experience is similar even in the presence of reasonable code abstractions like this. What are your thoughts on renaming `RoomFactory` to `VideoClient` and `newInstance` to `connect`.
> 
> Another option would be to implement an extension method off the `Video` class that allows connecting with a co-routine context.

That sounds good! I have no issues changing the name to `VideoClient` and the function to `connect`. @timrozum might want to be in sync on this as I'm pretty sure it's named `RoomFactory` in the iOS app.

Reply:
> Additionally, what are your thoughts on removing SharedPreferences from this class? It would be nice if this class did not know about shared preferences and simply accepted connect options as a method parameter. The call site would be responsible for getting the data for the connect options.

Hmm where do you imagine the `SharedPreferences` would be referenced? The whole point of this `RoomFactory` abstraction is to abstract the `ConnectOptions` and video connection code. This will leave the `RoomManager` with the sole responsibility of transmitting `RoomEvent`s to subscribers. It will also make unit testing the `RoomManager` much easier as you can easily mock out the `RoomFactory`.

Reply:
I guess I was thinking that the `Connect` event would contain the ConnectOptions required for connect. So whoever posts the Connect event would be responsible for reading shared preference data and creating the connect options.

Reply:
Ah I see. I like that idea! Yeah I'm not a fan of having the `SharedPreferences` as a dependency in the `RoomFactory`. I'll move it back to the activity layer and delegate the building of the `ConnectOptions` to another class.

Reply:
Actually the more I think about this, I don't think it is a good idea for the UI to be responsible for creating the `ConnectOptions`. I think it will complicate the UI and building the `ConnectOptions` would be part of the responsibility of the `RoomFactory`, which is to construct and connect to a room.

Reply:
Okay

Reply:
In iOS I do have a `RoomFactory` but it doesn't make a `TwilioVideo.Room` it makes a `VideoApp.Room`. The factory has a `makeRoom` function which is Swift naming convention and it really just makes the object. It is not kicking off work like fetching token or connecting.

`VideoApp.Room` manages the entire lifecycle of a room from just having a local participant (offline), fetching an access token, connecting, and disconnecting. 

`VideoApp.Room` uses `ConnectOptionsFactory` to make the connect options so that all that configuration is separate from lifecycle like connect and disconnect. 

In order to make UI and transform layers not a mess, in my SDK integration layer I had to define my own `Room`, `Participant`, `LocalParticipant`, `RemoteParticipant`, `VideoTrack`, `LocalVideoTrack`, and `RemoteVideoTrack`. My naming strategy for these types was to model the SDK and rely on Swift module namespaces to distinguish the two. Interface details model the SDK as much as possible so I'm thinking this matches the feedback from @aaalaniz above at least regarding actions. I did struggle with type names a bit but landed on just modeling the SDK. I'm not sure I could come up with new, intuitive names for all these types. I had to create these types for a number of reasons and can explain more if needed.

I do think that it is time to add an architecture section to the iOS readme to point customers in the right direction here regarding the different layers.

Reply:
Thanks for the response @timrozum ! I think one betterment here on the Android side to match what you have would be to create a `ConnectOptionsFactory` and add that as a dependency to the `VideoClient` to follow a more single responsibility approach. That way the `VideoClient` is just responsible for connecting to the room.